beacon-chain/blockchain/forkchoice/service.go

```diff
@@ -0,0 +1,460 @@
+//    )
+func (s *Store) GensisStore(genesisState *pb.BeaconState) error {
+	stateRoot, err := ssz.HashTreeRoot(genesisState)
+	if err != nil {
+		return errors.Wrap(err, "could not tree hash genesis state")
+	}
+
+	genesisBlk := blocks.NewGenesisBlock(stateRoot[:])
+
+	blkRoot, err := ssz.SigningRoot(genesisBlk)
+	if err != nil {
+		return errors.Wrap(err, "could not tree hash genesis block")
+	}
+
+	s.time = genesisState.GenesisTime
+	s.justifiedCheckpt = &ethpb.Checkpoint{Epoch: 0, Root: blkRoot[:]}
+	s.finalizedCheckpt = &ethpb.Checkpoint{Epoch: 0, Root: blkRoot[:]}
+
+	if err := s.db.SaveBlock(s.ctx, genesisBlk); err != nil {
+		return errors.Wrap(err, "could not save genesis block")
+	}
+	if err := s.db.SaveState(s.ctx, genesisState, blkRoot); err != nil {
+		return errors.Wrap(err, "could not save genesis state")
+	}
+
+	s.checkptBlkRootLock.Lock()
+	defer s.checkptBlkRootLock.Unlock()
+	h, err := hashutil.HashProto(s.justifiedCheckpt)
+	if err != nil {
+		return errors.Wrap(err, "could not hash proto justified checkpoint")
+	}
+	s.checkptBlkRoot[h] = blkRoot
+
+	return nil
+}
+
+// ancestor returns the block root of an ancestry block from the input block root.
+//
+// Spec pseudocode definition:
+//   def get_ancestor(store: Store, root: Hash, slot: Slot) -> Hash:
+//    block = store.blocks[root]
+//    assert block.slot >= slot
+//    return root if block.slot == slot else get_ancestor(store, block.parent_root, slot)
+func (s *Store) ancestor(root []byte, slot uint64) ([]byte, error) {
+	b, err := s.db.Block(s.ctx, bytesutil.ToBytes32(root))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get ancestor block")
+	}
+
+	// If we dont have the ancestor in the DB, simply return nil so rest of fork choice
+	// operation can proceed. This is not an error condition.
+	if b == nil || b.Slot < slot {
+		return nil, nil
+	}
+
+	if b.Slot == slot {
+		return root, nil
+	}
+
+	return s.ancestor(b.ParentRoot, slot)
+}
+
+// latestAttestingBalance returns the staked balance of a block from the input block root.
+//
+// Spec pseudocode definition:
+//   def get_latest_attesting_balance(store: Store, root: Hash) -> Gwei:
+//    state = store.checkpoint_states[store.justified_checkpoint]
+//    active_indices = get_active_validator_indices(state, get_current_epoch(state))
+//    return Gwei(sum(
+//        state.validators[i].effective_balance for i in active_indices
+//        if (i in store.latest_messages
+//            and get_ancestor(store, store.latest_messages[i].root, store.blocks[root].slot) == root)
+//    ))
+func (s *Store) latestAttestingBalance(root []byte) (uint64, error) {
+	s.checkptBlkRootLock.RLock()
+	defer s.checkptBlkRootLock.RUnlock()
+	h, err := hashutil.HashProto(s.justifiedCheckpt)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not hash proto justified checkpoint")
+	}
+	lastJustifiedBlkRoot := s.checkptBlkRoot[h]
+
+	lastJustifiedState, err := s.db.State(s.ctx, lastJustifiedBlkRoot)
```

Comment:
Do not use the Store's root context, require context in the arguments and use that